lib/mpool/lib/mblock_file.c

```diff
@@ -997,6 +1037,65 @@ mblock_file_delete(struct mblock_file *mbfp, uint64_t *mbidv, int mbidc)
     return 0;
 }
 
+merr_t
+mblock_punch(struct mblock_file *mbfp, uint64_t mbid, off_t off, size_t len)
+{
+    struct mblock_oid_info mbinfo;
+    uint32_t block;
+    bool prealloced;
+    size_t wlen;
+    merr_t err;
+    char *addr;
+    int rc;
+
+    INVARIANT(mbfp);
+
+    if (!PAGE_ALIGNED(off) || !PAGE_ALIGNED(len))
+        return merr(EINVAL);
+
+    wlen = mblock_wlen_get(mbfp, mbid);
+
+    if (off < 0 || off >= wlen || off + len > wlen)
+        return merr(EINVAL);
+
+    prealloced = mblock_is_prealloced(mbfp, mbid);
+    if (len == 0 || off + len == wlen) {
+        len = prealloced ? mbfp->mblocksz - off : wlen - off;
+        assert(PAGE_ALIGNED(len));
+    }
+
+    block = block_id(mbid);
+    err = mblock_rgn_find(&mbfp->rgnmap, block + 1);
+    if (err)
+        return err;
+
+    rc = fallocate(mbfp->fd, FALLOC_FL_PUNCH_HOLE | FALLOC_FL_KEEP_SIZE,
```

Comment:
When should one use fallocate vs posix_fallocate? Just a general question so that I can be smarter :)

Reply:
posix_fallocate(3) supports only pre-allocation of disk space, i.e., fallocate(2) with mode 0.

The nice thing about using posix_fallocate(3) vs fallocate(2) for pre-allocation is that posix_fallocate() provides an emulation if the underlying file-system doesn't support the fallocate operation in struct file_operations {}.